src/User.cpp

```diff
@@ -100,7 +100,57 @@ Watchable* RerunRecommenderUser::getRecommendation(Session& s) {
 GenreRecommenderUser::GenreRecommenderUser(const std::string& name): User(name), _popularTagsMap(),
                                                                      _popularTagsVector() {}
 
-GenreRecommenderUser::~GenreRecommenderUser() {}
+GenreRecommenderUser::~GenreRecommenderUser() {
+    this->cleanPopularTags(true);
+}
+
+void GenreRecommenderUser::cleanPopularTags(bool isDeepClean) {
+    for(uint i = 0; i < _popularTagsVector.size(); i++) {
+        PopularTag* tag = _popularTagsVector[i];
+        _popularTagsMap[tag->getName()] = nullptr;
+        _popularTagsVector[i] = nullptr;
+        if(isDeepClean)
+            delete tag;
+    }
+}
+
+void GenreRecommenderUser::deepCopyPopularTags(const std::vector<PopularTag*>& tags) {
+    for(auto tag : tags) {
+        PopularTag* newTag = new PopularTag(*tag);
+        _popularTagsVector.push_back(newTag);
+        _popularTagsMap[newTag->getName()] = newTag;
+    }
+}
+
+GenreRecommenderUser::GenreRecommenderUser(const GenreRecommenderUser& rhs) : User(rhs),
+                                                                              _popularTagsMap(),
+                                                                              _popularTagsVector() {
+    this->deepCopyPopularTags(rhs._popularTagsVector);
+}
+
+GenreRecommenderUser& GenreRecommenderUser::operator=(const GenreRecommenderUser& rhs) {
+    if(this != &rhs) {
+        this->cleanPopularTags(true);
+        this->deepCopyPopularTags(rhs._popularTagsVector);
+    }
+    return *this;
+}
+
+GenreRecommenderUser::GenreRecommenderUser(GenreRecommenderUser&& rhs) : User(rhs),
+                                                                _popularTagsMap(rhs._popularTagsMap),
+                                                                _popularTagsVector(rhs._popularTagsVector) {
+    rhs.cleanPopularTags(false);
```

Comment:
is it needed? isn't it happening thanks to vector::clear?

Reply:
Very needed, otherwise the destructor will deep clean